### PR TITLE
CRI: proposal for managing container stdout/stderr streams

### PR DESCRIPTION
**Pull request by yujuhong**
This scope of this proposal is limited to the stdout/stderr logs streams of the
containers.

This addresses #30709

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34376)

### CODE REVIEW

File: docs/proposals/kubelet-cri-logging.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+can easily extracts all necessary metadata (e.g., pod UID, container name) from
+the paths, and watch the directly for any changes.
+
+**Log format**
+
+The runtime should decorate each log entry with a RFC 3339 timestamp prefix,
+and ends with a newline.
+
+```
+2016-10-06T00:17:09.669794202Z The content of the log entry
```
Comment (feiskyer) [inline docs/proposals/kubelet-cri-logging.md]: Do we need to distinguish between `stdout` and `stderr`?

File: docs/proposals/kubelet-cri-logging.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+metadata associated with the log from the path. Note that this integration
+requires kubelet to know where the container runtime stores the logs, and will
+not be directly applicable to CRI.
+
+
+## Requirements
+
+
+   1. **Allow log collectors to integrate with kubernetes without knowledge of
+        the underlying container runtime.**
```
Comment (smarterclayton) [inline docs/proposals/kubelet-cri-logging.md]: Is this a requirement?   It's certainly nice to have, but I would have phrased this much weaker:

"Allow log collectors to integrate with Kubernetes across different container run times while preserving efficient storage and retrieval"

File: docs/proposals/kubelet-cri-logging.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+   2. **Allow kubelet to manage the lifecycle of the logs to pave the way for
+        better disk management in the future.** This implies that the lifecycle
+        of containers and their logs need to be decoupled.
+   3. **Provide ways for CRI-compliant runtimes to support all existing logging
```
Comment (smarterclayton) [inline docs/proposals/kubelet-cri-logging.md]: This would probably be #1 requirement in an ordered list.

File: docs/proposals/kubelet-cri-logging.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+defining complex logging API.
+
+There are also possibilities to implement the current proposal by imposing the
+log file paths, while leveraging the runtime to access and/or manage logs. This
+requires the runtime to expose knobs in CRI to retrieve, remove, and examine
+the disk usage of logs. The upside of this approach is that kubelet needs not
+mandate the logging format, assuming runtime already includes plugins for
+various logging formats. Unfortunately, this is not true for existing runtimes
+such as docker, which supports log retrieval only for a very limited number of
+log drivers [2]. On the other hand, the downside is that we would be enforcing
```
Comment (smarterclayton) [inline docs/proposals/kubelet-cri-logging.md]: Is this really a critical problem?  If supporting more log drivers in docker is important, we can always support that.

File: docs/proposals/kubelet-cri-logging.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+requirements, but no comprehensive solution for CRI specifically has been
+proposed so far (with the excpetion of @tmrtfs's proposal
+[#33111](https://github.com/kubernetes/kubernetes/pull/33111), which has a much
+wider scope). It has come up in discussions that kubelet can delegate all the
+logging management to the runtime to allow maximum flexibility. However, it is 
+difficult for this approach to meet either requirement (1) or (2), without
+defining complex logging API.
+
+There are also possibilities to implement the current proposal by imposing the
+log file paths, while leveraging the runtime to access and/or manage logs. This
```
Comment (smarterclayton) [inline docs/proposals/kubelet-cri-logging.md]: I'm not sure that this proposal makes a really strong argument against this (which I would consider the "default" path because it leverages the existing runtime support).

File: docs/proposals/kubelet-cri-logging.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+
+## Requirements
+
+
+   1. **Allow log collectors to integrate with kubernetes without knowledge of
+        the underlying container runtime.**
+
+   2. **Allow kubelet to manage the lifecycle of the logs to pave the way for
+        better disk management in the future.** This implies that the lifecycle
```
Comment (euank) [inline docs/proposals/kubelet-cri-logging.md]: This also implies the kubelet must be able to rotate logs, and thus must be able to work with the given format.

**Comment by coffeepac** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[docs/proposals/kubelet-cri-logging.md, line 35 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34376#-KTjRhpNf2O3nhMRHJ48:-KTjRhpNf2O3nhMRHJ49:bddxzzs) ([raw file](https://github.com/kubernetes/kubernetes/blob/18e7a0df8b5cbc6108cf29f429bf72daf79f7e10/docs/proposals/kubelet-cri-logging.md#L35)):*
> ```Markdown
> ## Goals and non-goals
> 
> Container Runtime Interfacej(CRI) is an ongoing project to allow container
> ```

small typo in "Interfacej(CRI)", there's a 'j' in there that probably shouldn't be.

---

*[docs/proposals/kubelet-cri-logging.md, line 116 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34376#-KTX2J-ImyZ9WwXYE93V-r1-116:-KTjTpmObrMihfehS_z3:b-z1wpmz) ([raw file](https://github.com/kubernetes/kubernetes/blob/18e7a0df8b5cbc6108cf29f429bf72daf79f7e10/docs/proposals/kubelet-cri-logging.md#L116)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> Is this a requirement?   It's certainly nice to have, but I would have phrased this much weaker:
> 
> 
> 
> "Allow log collectors to integrate with Kubernetes across different container run times while preserving efficient storage and retrieval"
</details>
I am in favor of the wording of the original requirement.  I'd like to be able to write a collector that interfaces with kubernetes, not 'kubernetes on docker' or 'kubernetes on rocket'.   I see where you're coming from in that you want to lock down the most important parts of the interface but I'd like the interface to not leak.

---

*[docs/proposals/kubelet-cri-logging.md, line 175 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34376#-KTX2J-ImyZ9WwXYE93V-r1-175:-KTjX6n3Bn1NRowEOn5J:bm1geky) ([raw file](https://github.com/kubernetes/kubernetes/blob/18e7a0df8b5cbc6108cf29f429bf72daf79f7e10/docs/proposals/kubelet-cri-logging.md#L175)):*
<details><summary><i>Previously, euank (Euan Kemp) wrote…</i></summary>
> For the kubectl logs feature it doesn't matter, for log aggregators it definitely could
</details>
For log aggregating it does matter.  Either separate files with stderr/stdout in the name or an additional field in the file of err/out.  To make 'kubectl logs' work easily, probably easier to add a field.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34376)*

File: docs/proposals/kubelet-cri-logging.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+when creating a container.
+
+```
+PodSandboxConfig.LogDirectory: /var/log/pods/<podUID>/
+ContainerConfig.LogPath: <containerName>_<instance#>.log
+```
+
+Because kubelet determines where the logs are stores and can access them
+directly, this meets requirement (1). As for requirement (2), the log collector
+can easily extracts all necessary metadata (e.g., pod UID, container name) from
```
Comment (jcantrill) [inline docs/proposals/kubelet-cri-logging.md]: ..all necessary metadata...   I would say this is misleading given for the openshift aggregated logging solution we consider the pod labels 'necessary' but are unable to 'access' them without a call to the api server.

Comment (euank) [inline docs/proposals/kubelet-cri-logging.md]: Agreed. Could we have a unique directory per pod (each being symlinked into that known path), where that directory contains a metadata file with the pod's name, id, annotations, and the same for its containers?

**Comment by euank** [comment]
cc @lucab, references your attach work as a path for rkt to integrate with this, and any insight you have would be appreciated!

I wouldn't be surprised if this comes up in sig-node tomorrow :)

File: docs/proposals/kubelet-cri-logging.md (hunk 5 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: +1 to @euank's comment, which explains more clearly than the statement in the proposal. 

TBH, I am not completely against leveraging runtime for various log operations. However, 
  1. We are not sure what we want yet, since disk management is still in its early stages. 
  2. Container runtimes such as docker also does not provide good support of logging yet. Even if we were to improve that (or hack docker internals in the shim), it'd take a longer time.

Instead of committing resources to do this right now, I think the current proposal gives us room to figure things out without having to develop and commit to a premature api.

Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: > We can question how much we actually need most of those requirements still though, since several of them are quite forward-looking.

I think I did this on purpose. I wanted to see if we can try finding a solution that addresses all requirements first. If, instead, sig-node thinks that some requirements can be dropped, it's perfectly reasonable to question the requirements themselves, and convince people who are in favor of the requirements.

File: docs/proposals/kubelet-cri-logging.md (hunk 4 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: This is just to point out that runtime support for log drivers are far from perfect today, and it's not good to simply say "runtime already supports log drivers, why can we reuse that?"

File: docs/proposals/kubelet-cri-logging.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+logging infrastructure directly to serve logs [1a]. Note that this does not
+eliminate the need to store the logs on the node locally for reliability.
+
+
+**How can existing runtimes (docker/rkt) comply to the logging requirements?**
+
+In the short term, the ongoing docker-CRI integration [3] will support the
+proposed solution only partially by (1) creating symbolic links for kubelet
+to access, but not manage the logs, and (2) add support for json format in
+kubelet. A more sophisticated solution that either involves using a custom plugin
```
Comment (yifan-gu) [inline docs/proposals/kubelet-cri-logging.md]: Is this saying for the near short term, we tolerate different logging formats than the one proposed above?

**Comment by yifan-gu** [comment]
Thank you @yujuhong :+1: 
I am ok with sticking to one logging format defined by kubelet if that's doable. 

@lucab @euank Does that mean we could make journald in rkt stage1 as optional?

**Comment by thockin** [comment]
I am OK with this proposal.  I understand all the arguments, and I think dithering any further is a waste of energy.  @yujuhong do you want to make more edits, or do you think it is good?  @smarterclayton are your questions advisory or a request to halt a merge?

**Comment by thockin** [comment]
wrt streaming through the runtime:

a) Can we get away with that FOR NOW?
b) Is it going to help us define and stabilize the overall design in the short term?
c) How expensive is it to pay off later?

File: docs/proposals/kubelet-cri-logging.md (hunk 1 above)
Comment (lucab) [inline docs/proposals/kubelet-cri-logging.md]: How is this supposed to interact with the container-app being attached to a TTY? It is my understanding that you can't actually get split stdout/stderr and a TTY at the same time, see https://github.com/docker/docker/issues/19696

File: docs/proposals/kubelet-cri-logging.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+or launching a separate process to copy and decorate the log will be considered
+as a mid-term solution.
+
+For rkt, there is a pending PR to enable systemd to redirect stdout/stderr to
+user-provided streams [4].
```
Comment (lucab) [inline docs/proposals/kubelet-cri-logging.md]: A bit more context on this:
```
For rkt, implementation will rely on providing external file-descriptors for stdout/stderr
to applications via systemd [4]. Those streams are currently managed by a journald
sidecar, which collects stream outputs and store them in the journal file of the pod.
This will replaced by a custom sidecar which can produce logs in the format expected
by this specification and can handle clients attaching as well.
```

**Comment by euank** [comment]
LGTM.

From the rkt side, we can implement this as @lucab commented, or by having some process act as a journald "remote" endpoint and have it transform logs to this format. The benefit of the latter is simply that the existing machined / journald integration would still work, but if you're using Kubernetes it's less important to have that and more important to have a generic on-node tools to work with pod logs (out of scope of this issue).

Until the implementation work is done rkt side, it's likely we'll want to have a runtime-specific workaround in kubelet.

File: docs/proposals/kubelet-cri-logging.md (hunk 6 above)
Comment (timothysc) [inline docs/proposals/kubelet-cri-logging.md]: I think this is an orthogonal concern... So long as it's not strict control.

File: docs/proposals/kubelet-cri-logging.md (hunk 2 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: Changed.

File: docs/proposals/kubelet-cri-logging.md (hunk 3 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: Done.

File: docs/proposals/kubelet-cri-logging.md (hunk 7 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: Changed the wording, and add the future work for writing a metadata file.

File: docs/proposals/kubelet-cri-logging.md (hunk 1 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: Added the stream type.

@lucab I think that's a limitation of the underlying implementation. For now, maybe we can do what docker does and report only stdout?

File: docs/proposals/kubelet-cri-logging.md (hunk 8 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: Yes. This helps bridge the transition.

File: docs/proposals/kubelet-cri-logging.md (hunk 9 above)
Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: Thanks for the explanation. Included this in the proposal.

File: docs/proposals/kubelet-cri-logging.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# CRI: Log management for container stdout/stderr streams
+
+
+## Goals and non-goals
+
+Container Runtime Interfacej(CRI) is an ongoing project to allow container
```
Comment (timstclair) [inline docs/proposals/kubelet-cri-logging.md]: nit: typo: `j`

**Comment by feiskyer** [comment]
Please run hack/update-munge-docs.sh

File: docs/proposals/kubelet-cri-logging.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+Because kubelet determines where the logs are stores and can access them
+directly, this meets requirement (1). As for requirement (2), the log collector
+can easily extract basic pod metadata (e.g., pod UID, container name) from
+the paths, and watch the directly for any changes. In the future, we can
+extend this by maintaining a metada file in the pod directory.
+
+**Log format**
+
+The runtime should decorate each log entry with a RFC 3339 timestamp prefix,
```
Comment (Random-Liu) [inline docs/proposals/kubelet-cri-logging.md]: s/RFC 3339/RFC 3339Nano

Found this during coding.
@yujuhong

File: docs/proposals/kubelet-cri-logging.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+the stream type (i.e., "stdout" or "stderr"), and ends with a newline.
```
Comment (euank) [inline docs/proposals/kubelet-cri-logging.md]: What about when the container was created with at tty?

In that case, there's no way to distinguish stdout from stderr (per discussion in 32869).

If you look at docker's json-log file for a container started with `-t`, it will only have `stdout` items.

Should we specify that here, or assume a `tty`'d container just shouldn't produce any logs (they'll often have escape sequences and such regardless)?

cc @lucab

Comment (Random-Liu) [inline docs/proposals/kubelet-cri-logging.md]: I think we may want to assume that container created with tty should only have logs with stream type "stdout".

Comment (euank) [inline docs/proposals/kubelet-cri-logging.md]: If we do, we should document this.

But I also think it could make sense to not support `logs` for that because a container with a tty is likely to emit terminal control characters or be used for read/writing large chunks of data (per @smarterclayton use-case for attach).

Comment (euank) [inline docs/proposals/kubelet-cri-logging.md]: Alternate option would be a special `tty` type btw. This could have value for log collectors being able to decide to ingress it or not, or for knowing when to play back as text to a terminal vs raw for any tool akin to `kubectl logs`

Comment (yujuhong) [inline docs/proposals/kubelet-cri-logging.md]: FYI, I created a new issue for this: #37247

**Comment by yujuhong** [comment]
@k8s-bot bazel test this

**Comment by yujuhong** [comment]
@k8s-bot e2e test this
